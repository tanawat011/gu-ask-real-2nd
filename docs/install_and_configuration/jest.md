# Jest for typescript

```text
`Jest` is a test runner.
```

## Documentation

- [Jest.io](https://jestjs.io/docs/getting-started)
- [testing-library](https://testing-library.com/docs/)

---

## Install and Configuration

### Install Jest for typescript

```sh
yarn add jest ts-jest
```

### Configuration

#### Set up test script in your [`package.json`](../package.json) file

```diff
...
 "scripts": {
    ...
+    "test": "react-scripts test --watchAll=false",
+    "test:w": "react-scripts test",
+    "test:cv": "yarn test --coverage",
+    "test:cv:msg": "yarn test:cv --verbose",
+    "test:cv:w": "yarn test:w --coverage",
+    "test:cv:msg:w": "yarn test:cv:w --verbose",
+    "test:cv:open": "open coverage/lcov-report/index.html",
    ...
  },
...
```

#### Set up basic configuration in your [`package.json`](../package.json) file

```diff
...
+  "jest": {
+    "coverageThreshold": {
+      "global": {
+        "statements": 40,
+        "branches": 40,
+        "functions": 40,
+        "lines": 40
+      }
+    },
+    "coveragePathIgnorePatterns": [
+      "src/pages/*",
+      "src/types/*",
+      "index.tsx",
+      "index.ts"
+    ]
+  }
...
```
