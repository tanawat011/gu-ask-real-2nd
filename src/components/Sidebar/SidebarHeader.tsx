import { useTranslation } from 'react-i18next'
import tw, { styled } from 'twin.macro'

import { TEXT_THEME } from 'constants/twTheme'

type SidebarHeaderProps = {
  fullSidebar: boolean
}

const TwContainer = styled.div(() => [
  tw`flex h-sidebar-header items-center justify-between px-6`,
])
const TwWrapLogo = tw.div`flex h-full items-center`
const TwLogo = tw.div`h-8 w-8 rounded-full bg-cadet-grey`
const TwTitle = styled.div(() => [[TEXT_THEME.MENU_HEAD], tw`ml-2 select-none`])

export const SidebarHeader: React.FC<SidebarHeaderProps> = ({
  fullSidebar,
}) => {
  const { t } = useTranslation('sidebar')

  return (
    <TwContainer>
      <TwWrapLogo>
        <TwLogo />

        {fullSidebar && <TwTitle>{t('header.title')}</TwTitle>}
      </TwWrapLogo>
    </TwContainer>
  )
}
