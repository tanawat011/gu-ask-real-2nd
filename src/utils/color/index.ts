export * from './hex2Rgb'
export * from './rgb2Hex'
