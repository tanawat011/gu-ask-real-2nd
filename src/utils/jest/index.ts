export * from './renderWithProviders'
export * from './twColor'
