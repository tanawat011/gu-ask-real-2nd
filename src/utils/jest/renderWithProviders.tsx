import type { RenderResult } from '@testing-library/react'

import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { RecoilRoot } from 'recoil'

const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  const queryClient = new QueryClient()

  return (
    <QueryClientProvider client={queryClient}>
      <RecoilRoot>
        <BrowserRouter>{children}</BrowserRouter>
      </RecoilRoot>
    </QueryClientProvider>
  )
}

interface RenderWithProvidersOptions {
  router?: {
    route: string
  }
  recoil?: object
}

export const renderWithProviders = (
  ui: JSX.Element,
  option?: RenderWithProvidersOptions,
): RenderResult => {
  const routerOption = { ...(option?.router || { route: '/' }) }
  window.history.pushState({}, 'Test page', routerOption.route)

  return {
    ...render(ui, { wrapper: AllTheProviders, ...routerOption }),
  }
}
