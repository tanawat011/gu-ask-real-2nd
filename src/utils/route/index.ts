export * from './menuGenerator'
export * from './renderRoutes'
export * from './routeGenerator'
