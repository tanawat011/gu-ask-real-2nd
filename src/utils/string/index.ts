export * from './camelToKebab'
export * from './camelToSentence'
export * from './replaceAll'
