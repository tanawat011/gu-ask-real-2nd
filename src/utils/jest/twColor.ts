import { hex2Rgb } from 'utils/color'

export const twColor = (hexColor: string, type?: 'bg' | 'border' | 'text') => {
  if (hexColor === 'transparent') {
    return 'transparent'
  }

  if (hexColor.length > 7) {
    return 'undefined'
  }

  const rgb = hex2Rgb(hexColor, { isArray: true }) as string[]

  if (!type) {
    type = 'bg'
  }

  return `rgb(${rgb.join(' ')} / var(--tw-${type}-opacity))`
}
