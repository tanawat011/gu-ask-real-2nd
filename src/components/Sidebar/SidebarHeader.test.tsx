import { renderWithProviders } from 'utils/jest'

import { SidebarHeader } from './SidebarHeader'

describe('<SidebarHeader />', () => {
  test('should render without crashing and match snapshot', () => {
    expect.hasAssertions()

    const { asFragment } = renderWithProviders(<SidebarHeader fullSidebar />)

    expect(asFragment()).toMatchSnapshot()
  })
})
