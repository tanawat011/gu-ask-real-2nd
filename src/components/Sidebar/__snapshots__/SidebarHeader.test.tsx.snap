// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarHeader /> should render without crashing and match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 64px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding-left: 1.5rem;
  padding-right: 1.5rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  height: 2rem;
  width: 2rem;
  border-radius: 9999px;
  --tw-bg-opacity: 1;
  background-color: rgb(156 163 175 / var(--tw-bg-opacity));
}

.c4 {
  --tw-text-opacity: 1;
  color: rgb(107 114 128 / var(--tw-text-opacity));
  margin-left: 0.5rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.dark .c3 {
  --tw-text-opacity: 1;
  color: rgb(156 163 175 / var(--tw-text-opacity));
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      />
      <div
        class="c3 c4"
      >
        Gu Ask Real
      </div>
    </div>
  </div>
</DocumentFragment>
`;
